lib/shared-method.js

```diff
@@ -255,8 +255,41 @@ function coerceAccepts(uarg, desc) {
 
   var actualType = SharedMethod.getType(uarg);
 
-  // is the arg optional?
-  // arg was not provided
+  // convert values to the correct type
+  // TODO(bajtos) Move conversions to HttpContext (and friends)
+  // SharedMethod should only check that argument values match argument types.
+  var conversionNeeded = targetType !== 'any' &&
+    actualType !== 'undefined' &&
+    actualType !== targetType;
+
+  if (conversionNeeded) {
+    // JSON.parse can throw, so catch this error.
+    try {
+      uarg = convertValueToTargetType(name, uarg, targetType);
+      actualType = SharedMethod.getType(uarg);
+    } catch (e) {
+      var message = util.format('invalid value for argument \'%s\' of type ' +
+        '\'%s\': %s. Received type was %s. Error: %s',
+        name, targetType, uarg, typeof uarg, e.message);
+      throw new badArgumentError(message);
+    }
+  }
+
+  var typeMismatch = targetType !== 'any' &&
+    actualType !== 'undefined' &&
+    targetType !== actualType &&
+    // an array can be assigned to an object
+    !(targetType === 'object' && actualType === 'array');
```

Comment:
When you have an argument of `type: 'object'`, this line allows to assign an array value to this argument. In JavaScript, an array is an object too (`typeof [] === 'object'`). However, `SharedMethod.getType([])` returns `array` instead of `object`.

Does it make sense now? Should I include the whole paragraph above in the comment?